design/complex-workloads.md

```diff
@@ -0,0 +1,988 @@
+      reason: ReplicaSetUpdated
+      status: "True"
+      type: Progressing
+```
+
+Listing components associated with a Kubernetes application:
+```bash
+$ kubectl --context crossplane get kubernetesapplication demo
+NAME  DESIRED SUBMITTED  CLUSTER                      AGE
+demo  1       1          theperfectkubernetescluster  8s
+
+$ kubectl --context crossplane get kubernetesapplicationcomponent --selector app=demo
+NAME  KIND        CLUSTER                      STATUS     AGE
+demo  Deployment  theperfectkubernetescluster  Submitted  8s
+```
+
+#### Naming
+The proposed `KubernetesApplication` and especially
+`KubernetesApplicationComponent` names are rather verbose when compared to their
+contemporary: `Workload`. These names are best justified by breaking them down
+into their parts:
+
+_Kubernetes_ represents the deployment vector of the application. Prefixing the
+kind with Kubernetes leaves room to define applications that are deployed using
+other methods. This design proposes the explicit prefix Kubernetes rather than
+the abstract prefix _Containerized_ because the proposed CRD is tightly coupled
+to Kubernetes; it could not be used to deploy a containerized application via
+Amazon ECS or Docker Swarm. The scheme chosen by this design impacts future
+implementations; would an application targeting Amazon Lambda be named a
+ServerlessApplication or a LambdaApplication? Kubernetes is arguably ubiquitous
+enough to be analogous with generic resource kind names like
+ServerlessApplication or VMApplication.
+
+_Application_ distinguishes a workload from a compute resource when interacting
+with Crossplane. It is synonymous in this context with _Workload_, which is
+implied by the workload.crossplane.io API namespace. Including Application would
+thus be redundant except that the API namespace is typically omitted when
+interacting with the API server. Assume `KubernetesApplication` was instead
+named `Kubernetes`, relying on the API namespace to indicate that it was a
+workload. In this scenario `kubectl get kubernetes` would return Kubernetes
+workloads while `kubectl get kubernetescluster` would return Crossplane managed
+Kubernetes clusters. These names are close enough that it's not unlikely
+Crossplane users would expect `kubectl get kubernetes` to return Kubernetes
+clusters rather than workloads. Application is preferable to Workload to avoid
+[stuttering][5] when the API namespace is considered, and provides symmetry with
+similar concepts like sig-apps' [Application][6].
+
+_Component_ represents an arbitrary resource of which an application consists.
+A component could template a compute resource such as a `Deployment`,
+`StatefulSet`, or `Job`; a configuration resource such as a `ConfigMap` or
+`Secret`; or a networking resource such as a `Service` or `Ingress`. The word
+Component, defined as "a part or element of a larger whole" suits this generic
+case.
```

Comment:
_Resource_ may also be a good name here. Kubernetes seems to use _resource_ and _object_ somewhat interchangeably when referring to "a thing you model in the API server that has type and object metadata".

If you consider Deployment, Service, ConfigMap, etc to be resources then it would make sense that a KubernetesApplicationResource contains a template of a resource.

Reply:
my initial reaction was hesitation because of how overloaded of a term `Resource` seems to be in the k8s world (and Crossplane world too).  But this does actually seem to be an accurate usage of the term in an accurate k8s context, so I'm not against it :)

Reply:
For the record I've run with `KubernetesApplicationResource` in the branch in which I'm working on controller implementations. It will be easy enough to rename if we change our minds.